### PR TITLE
Add option to run actions sequentially.

### PR DESCRIPTION
**Pull request by dompuiu**

### CODE REVIEW

**Comment by dompuiu** [comment]
It seems SauceLabs has some problems. The jobs are failing from timeouts.

File: src/__tests__/initRules.test.js [hunk 1]
```diff
@@ -13,639 +13,2329 @@
...
 
-      var action1Exports = moduleProvider.getModuleExports(TEST_ACTION1_PATH);
+      it('executes the events then the conditions then the actions', function() {
+        var callOrder = [];
+        var rules = setupRules([
+          {
+            events: [
+              generateEvent('Event1', function(module) {
+                callOrder.push('Event1');
+                module.exports = jasmine
```
Comment (Aaronius) [inline src/__tests__/initRules.test.js]: I prefer we just use regular functions unless we're asserting against the spies.

File: src/__tests__/initRules.test.js [hunk 2]
```diff
@@ -13,639 +13,2329 @@
...
+        );
+      });
 
-      var errorMessage = logger.error.calls.mostRecent().args[0];
-      expect(errorMessage).toBe('Failed to execute ' + TEST_CONDITION1_DISPLAY_NAME +
-        ' for Test Rule rule. Module did not export a function.');
-      expect(notifyMonitors).toHaveBeenCalledWith('ruleConditionFailed', {
-        rule: rules[0],
-        condition: rules[0].conditions[0]
+      it('does not throw error when there are no conditions for a rule
```
Comment (Aaronius) [inline src/__tests__/initRules.test.js]: Typo: a condition module

This is in subsequent tests as well.

Comment (Aaronius) [inline src/__tests__/initRules.test.js]: Typo: future

File: src/__tests__/initRules.test.js [hunk 3]
```diff
@@ -13,639 +13,2328 @@
...
+        );
+      });
 
-      var errorMessage = logger.error.calls.mostRecent().args[0];
-      expect(errorMessage).toBe('Failed to execute ' + TEST_CONDITION1_DISPLAY_NAME +
-        ' for Test Rule rule. Module did not export a function.');
-      expect(notifyMonitors).toHaveBeenCalledWith('ruleConditionFailed', {
-        rule: rules[0],
-        condition: rules[0].conditions[0]
+      it('does not throw err
```
[the remark this answers is not in the document]
Comment (Aaronius) [reply src/__tests__/initRules.test.js]: I think we need tests for the timeouts. When the timeouts reject the promise, the error message should probably say something about how the condition or action took too long to execute.

File: src/initRules.js [hunk 4]
```diff
@@ -24,166 +28,255 @@ module.exports = function(
...
+      rule.conditions.forEach(function(condition) {
+        rulesQueue.lastPromiseInQueue = rulesQueue.lastPromiseInQueue.then(
+          function() {
+            var conditionPromise = new Promise(function(resolve) {
+              resolve(
+                executeDelegateModule(condition, syntheticEvent, [
+                  syntheticEvent
+                ])
+              );
+            })
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/initRules.js]: Errors thrown inside the module, would not be caught by the catch clause.

File: src/initRules.js [hunk 5]
```diff
@@ -24,166 +28,255 @@ module.exports = function(
...
+              .catch(function(e) {
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/initRules.js]: Prettier does this format. :/

File: src/createRulesPromiseQueue.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * governing permissions and limitations under the License.
+ ****************************************************************************************/
+
+var Promise = require('@adobe/reactor-promise');
+
+module.exports = function() {
+  return {
+    lastPromiseInQueue: Promise.resolve()
+  };
+};
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/createRulesPromiseQueue.js]: Done what we've talked in person.

File: src/__tests__/initRules.test.js (hunk 2 above)
Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Done.

File: src/__tests__/helpers/module.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    if (!str) return '';
+
+    return String(str)
+      .replace(/^[^A-Za-z0-9]*|[^A-Za-z0-9]*$/g, '')
+      .replace(/([a-z])([A-Z])/g, function(m, a, b) {
+        return a + '-' + b.toLowerCase();
+      })
+      .replace(/[^A-Za-z0-9]+|_-+/g, '-')
+      .toLowerCase();
+  },
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/__tests__/helpers/module.js]: Done.

File: src/__tests__/initRules.test.js [hunk 8]
```diff
@@ -13,639 +13,2328 @@
 'use strict';
 var createModuleProvider = require('../createModuleProvider');
 var injectInitRules = require('inject-loader!../initRules');
+var ModuleHelper = require('./helpers/module');
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Done.

File: src/__tests__/initRules.test.js (hunk 1 above)
Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Done.

File: src/__tests__/initRules.test.js [hunk 9]
```diff
@@ -13,639 +13,2328 @@
...
+          }
+        ]);
+
+        runInitRules(rules);
 
-      var condition1Exports = moduleProvider.getModuleExports(TEST_CONDITION1_PATH);
+        expect(callOrder).toEqual(['Action1', 'Action2']);
+      });
+
+      it('ceases to execute remaining conditions when condition fails', function() {
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Done.

File: src/__tests__/initRules.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Done.

File: src/__tests__/initRules.test.js [hunk 10]
```diff
@@ -13,639 +13,2328 @@
...
+        );
+        var conditionExport2 = moduleProvider.getModuleExports(
+          ModuleHelper.getPath('Condition2')
+        );
+
+        expect(conditionExport1.calls.count()).toBe(1);
+        expect(conditionExport2.calls.count()).toBe(1);
+      });
+
+      it('executes all the rule conditions in the order provided', function() {
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Done.

**Comment by Aaronius** [comment]
Now that failed actions can prevent a rule from completing, I think we need to notify monitors in `logActionError` similar to how we do so in `logConditionError`.

File: src/__tests__/initRules.test.js [hunk 11]
```diff
@@ -510,6 +513,35 @@ describe('initRules', function() {
...
+          moduleHelper.getPath('Action1')
+        );
+        var action2Export = moduleProvider.getModuleExports(
+          moduleHelper.getPath('Action1')
+        );
+
+        expect(action1Export.calls.count()).toBe(1);
+        expect(action2Export.calls.count()).toBe(1);
+      });
+
```
Comment (Aaronius) [inline src/__tests__/initRules.test.js]: I think we need the same test for when the queue is not being used.

Comment (Aaronius) [reply src/__tests__/initRules.test.js]: Oh nevermind. I misunderstood.

Comment (Aaronius) [reply src/__tests__/initRules.test.js]: I think I means...we need the same test for when the queue IS being used.

File: src/__tests__/initRules.test.js [hunk 12]
```diff
@@ -1584,13 +1598,11 @@ describe('initRules', function() {
         'does execute the next action in the chain if any action takes' +
           ' longer than 2 seconds to complete',
```
Comment (Aaronius) [inline src/__tests__/initRules.test.js]: I think this test should say "does not execute" now.

Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Done.

File: src/__tests__/initRules.test.js (hunk 11 above)
Comment (dompuiu) [reply src/__tests__/initRules.test.js]: Added a test for the queue set. FIxed also this test which was wrong.

**Pull request merged**